From 129adfab49a4db739245796fd731d6cda4e33d69 Mon Sep 17 00:00:00 2001
From: Vipul Singh <imvipul@google.com>
Date: Wed, 6 May 2026 11:06:19 -0700
Subject: [PATCH] Remove usage of Parcel.allowSquashing in RemoteViews and
 immediate unparceling of Bundle

Squashing was added for application info but it doesn't work so mApplicationInfoCache was also added. Removing this from RemoteView as its redundant. And removed early unparceling of Bundle in ReflectionAction.

Flag: EXEMPT BUGFIX
Bug: 477278752
Test: m sts && sts-tradefed run sts-dynamic-develop -m StsComposedHostTestCases --compatibility:test-arg "com.android.compatibility.common.tradefed.testtype.JarHostTest:include-annotation:android.platform.test.annotations.StsPresubmit"
Cherrypick-From: https://googleplex-android-review.googlesource.com/q/commit:e1b25527c312b0d87b5399ca383d2864c9767738
Merged-In: Ie717bcaa9a86f998836144f216318cb9addbae0d
Change-Id: Ie717bcaa9a86f998836144f216318cb9addbae0d
---
 core/java/android/widget/RemoteViews.java | 22 +---------------------
 1 file changed, 1 insertion(+), 21 deletions(-)

diff --git a/core/java/android/widget/RemoteViews.java b/core/java/android/widget/RemoteViews.java
index 1e0c4906e6ed..ccd15778bfe0 100644
--- a/core/java/android/widget/RemoteViews.java
+++ b/core/java/android/widget/RemoteViews.java
@@ -344,13 +344,6 @@ public class RemoteViews implements Parcelable, Filter {
     static final int FLAG_MASK_TO_PROPAGATE =
             FLAG_WIDGET_IS_COLLECTION_CHILD | FLAG_USE_LIGHT_BACKGROUND_LAYOUT;
 
-    /**
-     * A ReadWriteHelper which has the same behavior as ReadWriteHelper.DEFAULT, but which is
-     * intentionally a different instance in order to trick Bundle reader so that it doesn't allow
-     * lazy initialization.
-     */
-    private static final Parcel.ReadWriteHelper ALTERNATIVE_DEFAULT = new Parcel.ReadWriteHelper();
-
     /**
      * Used to restrict the views which can be inflated
      *
@@ -2894,18 +2887,7 @@ public class RemoteViews implements Parcelable, Filter {
                     this.mValue = in.readTypedObject(Bitmap.CREATOR);
                     break;
                 case BUNDLE:
-                    // Because we use Parcel.allowSquashing() when writing, and that affects
-                    //  how the contents of Bundles are written, we need to ensure the bundle is
-                    //  unparceled immediately, not lazily.  Setting a custom ReadWriteHelper
-                    //  just happens to have that effect on Bundle.readFromParcel().
-                    // TODO(b/212731590): build this state tracking into Bundle
-                    if (in.hasReadWriteHelper()) {
-                        this.mValue = in.readBundle();
-                    } else {
-                        in.setReadWriteHelper(ALTERNATIVE_DEFAULT);
-                        this.mValue = in.readBundle();
-                        in.setReadWriteHelper(null);
-                    }
+                    this.mValue = in.readBundle();
                     break;
                 case INTENT:
                     this.mValue = in.readTypedObject(Intent.CREATOR);
@@ -8926,7 +8908,6 @@ public class RemoteViews implements Parcelable, Filter {
 
     private void writeToParcel(Parcel dest, int flags,
             @Nullable SparseArray<Intent> intentsToIgnore) {
-        boolean prevSquashingAllowed = dest.allowSquashing();
 
         if (!hasMultipleLayouts()) {
             dest.writeInt(MODE_NORMAL);
@@ -8973,7 +8954,6 @@ public class RemoteViews implements Parcelable, Filter {
         dest.writeLong(mProviderInstanceId);
         dest.writeBoolean(mHasDrawInstructions);
 
-        dest.restoreAllowSquashing(prevSquashingAllowed);
     }
 
     private void writeActionsToParcel(Parcel parcel, int flags) {
-- 
2.55.0.897.gb25b4bd76c-goog

